File: 3_Structural_Design_Patterns/3.03_Composite/Python_Snippet/Composite.py
from abc import ABC, abstractmethod

class FileSystem(ABC):
    @abstractmethod
    def display_info(self):
        pass

class File(FileSystem):
    def __init__(self, name, contents):
        self.name = name
        self.contents = contents

    def display_info(self):
        print(f"File: {self.name}, Contents: {self.contents}")

class Directory(FileSystem):
    def __init__(self, name):
        self.name = name
        self.children = []

    def add_child(self, child):
        self.children.append(child)

    def display_info(self):
        print(f"Directory: {self.name}")
        for child in self.children:
            child.display_info()

# Usage
if __name__ == "__main__":
    base_dir = Directory("home")
    hack_file = File("hack.txt", "Hacking NASA with HTML")
    hack_dir = Directory("hack")
    html_file = File("index.html", "<h1>Welcome to NASA</h1>")
    base_dir.add_child(html_file)
    base_dir.add_child(hack_dir)
    hack_dir.add_child(hack_file)
    base_dir.display_info()

File: 3_Structural_Design_Patterns/3.03_Composite/Java_Snippet/Composite.java
import java.util.*;

interface FileSystem {
    void displayInfo();
}

class File implements FileSystem {
    private String name;
    private String contents;

    public File(String name, String contents) {
        this.name = name;
        this.contents = contents;
    }

    @Override
    public void displayInfo() {
        System.out.println("File: " + name + ", Contents: " + contents);
    }
}

class Directory implements FileSystem {
    private String name;
    private List<FileSystem> children;

    public Directory(String name) {
        this.name = name;
        this.children = new ArrayList<>();
    }

    public void addChild(FileSystem child) {
        this.children.add(child);
    }

    @Override
    public void displayInfo() {
        System.out.println("Directory: " + this.name);
        for (FileSystem child : this.children) {
            child.displayInfo();
        }
    }
}

public class Composite {
    public static void main(String[] args) {
        Directory baseDir = new Directory("home");
        File hackFile = new File("hack.txt", "Hacking NASA with HTML");
        Directory hackDir = new Directory("hack");
        File htmlFile = new File("index.html", "<h1>Welcome to NASA</h1>");
        baseDir.addChild(htmlFile);
        baseDir.addChild(hackDir);
        hackDir.addChild(hackFile);
        baseDir.displayInfo();
    }
}